pangeo_forge_recipes/recipes/xarray_zarr.py

```diff
@@ -46,6 +48,480 @@ def _chunk_metadata_fname(chunk_key) -> str:
 ChunkKey = Tuple[int]
 InputKey = Tuple[int]
 
+
+def expand_target_dim(target: CacheFSSpecTarget, concat_dim: Optional[str], dimsize: int) -> None:
+    target_mapper = target.get_mapper()
+    zgroup = zarr.open_group(target_mapper)
+    ds = open_target(target)
+    sequence_axes = {
+        v: ds[v].get_axis_num(concat_dim) for v in ds.variables if concat_dim in ds[v].dims
+    }
+
+    for v, axis in sequence_axes.items():
+        arr = zgroup[v]
+        shape = list(arr.shape)
+        shape[axis] = dimsize
+        logger.debug(f"resizing array {v} to shape {shape}")
+        arr.resize(shape)
+
+    # now explicity write the sequence coordinate to avoid missing data
+    # when reopening
+    if concat_dim in zgroup:
+        zgroup[concat_dim][:] = 0
+
+
+def open_target(target: CacheFSSpecTarget) -> xr.Dataset:
+    return xr.open_zarr(target.get_mapper())
+
+
+def input_position(input_key, file_pattern: FilePattern, concat_dim: Optional[str]):
+    assert concat_dim is not None
+    concat_dim_axis = list(file_pattern.dims).index(concat_dim)
+    return input_key[concat_dim_axis]
+
+
+def cache_input_metadata(
+    input_key: InputKey,
+    metadata_cache: Optional[MetadataTarget],
+    file_pattern: FilePattern,
+    input_cache: Optional[CacheFSSpecTarget],
+    cache_inputs: bool,
+    copy_input_to_local_file: bool,
+    xarray_open_kwargs: dict,
+    delete_input_encoding: bool,
+    process_input: Optional[Callable[[xr.Dataset, str], xr.Dataset]],
+):
+    # TODO(TOM): figure out where caching should happen
+    if metadata_cache is None:
+        raise ValueError("metadata_cache is not set.")
+    logger.info(f"Caching metadata for input '{input_key}'")
+    with open_input(
+        input_key,
+        file_pattern=file_pattern,
+        input_cache=input_cache,
+        cache_inputs=cache_inputs,
+        copy_input_to_local_file=copy_input_to_local_file,
+        xarray_open_kwargs=xarray_open_kwargs,
+        delete_input_encoding=delete_input_encoding,
+        process_input=process_input,
+    ) as ds:
+        input_metadata = ds.to_dict(data=False)
+        # TODO(METADATA): set
+        metadata_cache[_input_metadata_fname(input_key)] = input_metadata
+
+
+def cache_input(
+    input_key: InputKey,
+    cache_inputs: bool,
+    input_cache: Optional[CacheFSSpecTarget],
+    file_pattern: FilePattern,
+    fsspec_open_kwargs: dict,
+    cache_metadata: bool,
+    copy_input_to_local_file: bool,
+    xarray_open_kwargs: dict,
+    delete_input_encoding: bool,
+    process_input: Optional[Callable[[xr.Dataset, str], xr.Dataset]],
+    metadata_cache: Optional[MetadataTarget],
```

Comment:
I feel like these long argument blocks may hurt maintainability. There is so much room for programmer error when passing long lists of arguments through the stack. If we go this route, perhaps we want to enclose some of the arguments in a `Config` object?

Reply:
Agreed! Though I *think* the biggest risk of programmer error is right now / in future major refactors. So it comes down to simplicity of the code (not having to look up what is this "FooConfig" thing?) vs. typing all these out. I'm 51% in favor of writing things out explicitly like this, but am happy to go with a Config / Options style object.

I will say that mypy has been helpful here. It caught a couple issues before running the tests.